### DISCUSSION TITLE
Sqlalchemy query indefinitely hangs after a internet reconnection

### DISCUSSION THREAD

**Discussion by ZhimaoLin** [sqlalchemy/sqlalchemy#12009] [Usage Questions]
We are trying to run a database migration script using Sqlalchemy. When the script is running, if I disconnect the internet and reconnect the internet, the script hangs indefinitely. There is no error or exception raised. I am expecting Sqlalchemy to raise an error such as connection error or timeout error. However, if I completely shut down the internet and NOT connect it back, it raises an connection error as expected. 

Sqlalchemy version: `2.0.36`
Database driver: `mysql-connector-python 8.3.0`
DB vendor: `MySQL 8.0.34`
Python version: `3.11.6`
Operating system: `MacOS Sequoia 15.0.1`

Here is an overview of my Sqlalchemy code:
```
engine = create_engine(DB_URL)
Session = sessionmaker(autocommit=False, bind=engine)
db = Session()

try:
    for i in range(5000): # Long loop takes about 10 minutes
        # Query some data in the DB
        db.query(xxx).filter(xxx).all()

        # Modify some data in the DB
        db.query(xxx).filter(xxx).update(xxx)

        # Add some data to the DB
        db.add(xxx)

        # Commit the transaction
        db.commit()
except Exception as e:
    print(e)
    db.rollback()
finally:
    db.close()
```
When the script hangs, I cannot see any error print on my Terminal.

Can someone tell me why Sqlalchemy hangs after the internet reconnection and any solution to this issue? 
Thank you!

**Comment by zzzeek** [comment]
this is a driver issue and there should be a timeout that you can set for the driver.   I'm not too confident on mysql-connector-python however if you use mysqlclient you can set timeouts, and you might also want to see if a change in driver just fixes the problem since we've had many problems with mysql-connector-python.  you want to look at [read_timeout and write_timeout](https://dev.mysql.com/doc/c-api/8.0/en/mysql-options.html)
